c/shared/source/uforead/uforead.c

```diff
@@ -1819,9 +1818,153 @@ static void reallocFDArray(ufoCtx h){
     h->top.FDArray.array = newFDArray;
 }
 
+static void parseXMLFile(ufoCtx h, char* filename, const char* filetype){
+    xmlDocPtr doc;
+    xmlNodePtr cur;
+    unsigned char* keyID;
+    char* keyName;
+
+    xmlKeepBlanksDefault(0);
+
+    h->cb.stm.xml_read(&h->cb.stm, h->stm.src, &h->src.buf, &doc);
+
+    if (doc == NULL) {
+        fatal(h, ufoErrParse, "Failed to parse '%s'.\n", "file");
+    }
+
+    cur = xmlDocGetRootElement(doc);
+    if (cur == NULL) {  // if document empty
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    if (!xmlStrEqual((cur)->name, (const xmlChar *) filetype)) {
+        fprintf(stderr, "document of the wrong type, root node != %s", filetype);
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    cur = (cur)->xmlChildrenNode;
+    while (cur && xmlIsBlankNode(cur)) {
+        cur = (cur) -> next;
+    }
+    if ( cur == 0 ) {
+        xmlFreeDoc(doc);
+//        return ( NULL );
+    }
+    if ((!xmlStrEqual((cur)->name, (const xmlChar *) "dict"))) {
+        fprintf(stderr, "Error reading outermost <dict> in %s.\n", "file");
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        keyID = NULL;
+        if ((xmlStrEqual(cur->name, (const xmlChar *) "key"))) {
+            parseKeyName(&keyID, cur);
+            if (keyID == NULL){
+                message(h, "Warning: Encountered empty <key></key>.");
+                cur = cur->next;
+                continue;
+            }
+            keyName = (char*) keyID;
+            cur = cur->next;
```

Comment:
This could well be right but could you briefly explain why we're advancing twice when we read a key with a parseable name and once when it's not a key or we can't parse the name?

Reply:
When we read a nonekey or an 'empty' key (one that doesn't have a value defined for it), that one cur = cur -> next advances the document ptr to the next item (which will be the empty key value, or the next key if that key value was never defined). This value is then advanced over because it isn't a key, thus advancing again. So technically, it is also advanced twice.

When reading a key that are parseable: I set up the advancing so that I would call the parse functions, but still have the ptr at the same spot in case I wanted to do other things with it. To me, it seems like the caller should have ptr management responsibility. Do you think that makes sense?

Reply:
@skef says "I understand the basic idea here now. You'll check the potential case of loosing the second key of two keys in a row."

Reply:
@skef says "I understand the basic idea here now. You'll check the potential case of loosing the second key of two keys in a row."

Reply:
I've refactored this part of the code (`parseXMLKeyName` and its callers) to clarify how `cur` is being advanced. You'll now see that the `keyName` validation happens within `parseXMLKeyName`.

I've also added tests to `tx_test.py::test_ufo_fontinfo_parsing` for cases like this, such as two keys back to back (with a value missing for the first one) to make sure we don't skip over the next key.

Based on those points, I think this can be resolved unless you have additional thoughts